ethcore/engines/authority-round/src/randomness.rs

```diff
@@ -0,0 +1,257 @@
+	/// Determine randomness generation state from the contract.
+	///
+	/// Calls various constant contract functions to determine the precise state that needs to be
+	/// handled (that is, the phase and whether or not the current validator still needs to send
+	/// commitments or reveal random numbers).
+	pub fn load(
+		contract: &BoundContract,
+		our_address: Address,
+	) -> Result<RandomnessPhase, PhaseError> {
+		// Determine the current round and which phase we are in.
+		let round = contract
+			.call_const(aura_random::functions::current_collect_round::call())
+			.map_err(PhaseError::LoadFailed)?;
+		let is_commit_phase = contract
+			.call_const(aura_random::functions::is_commit_phase::call())
+			.map_err(PhaseError::LoadFailed)?;
+
+		// Ensure we are not committing or revealing twice.
+		let committed = contract
+			.call_const(aura_random::functions::is_committed::call(
+				round,
+				our_address,
+			))
+			.map_err(PhaseError::LoadFailed)?;
+		let revealed: bool = contract
+			.call_const(aura_random::functions::sent_reveal::call(
+				round,
+				our_address,
+			))
+			.map_err(PhaseError::LoadFailed)?;
+
+		// With all the information known, we can determine the actual state we are in.
+		if is_commit_phase {
+			if revealed {
+				return Err(PhaseError::RevealedInCommit);
+			}
+
+			if !committed {
+				Ok(RandomnessPhase::BeforeCommit)
+			} else {
+				Ok(RandomnessPhase::Committed)
+			}
+		} else {
+			if !committed {
+				// We apparently entered too late to make a commitment, wait until we get a chance again.
+				return Ok(RandomnessPhase::Waiting);
+			}
+
+			if !revealed {
+				Ok(RandomnessPhase::Reveal { our_address, round })
+			} else {
+				Ok(RandomnessPhase::Waiting)
+			}
+		}
+	}
+
+	/// Advance the random seed construction process as far as possible.
+	///
+	/// Returns the encoded contract call necessary to advance the randomness contract's state.
+	///
+	/// **Warning**: After calling the `advance()` function, wait until the returned transaction has been included in
+	/// a block before calling it again; otherwise spurious transactions resulting in punishments might be executed.
+	pub fn advance<R: Rng>(
+		self,
+		contract: &BoundContract,
+		rng: &mut R,
+		signer: &dyn EngineSigner,
+	) -> Result<Option<Bytes>, PhaseError> {
+		match self {
+			RandomnessPhase::Waiting | RandomnessPhase::Committed => Ok(None),
+			RandomnessPhase::BeforeCommit => {
+				// Generate a new random number, but don't reveal it yet. Instead, we publish its hash to the
+				// randomness contract, together with the number encrypted to ourselves. That way we will later be
+				// able to decrypt and reveal it, and other parties are able to verify it against the hash.
+				let number: RandNumber = rng.gen();
+				let number_hash: Hash = keccak(number.as_bytes());
+				let public = signer.public().ok_or(PhaseError::MissingPublicKey)?;
+				let cipher = ecies::encrypt(&public, &number_hash.0, number.as_bytes())?;
```

Comment:
We think it would be more reliable to store the cipher on-chain, and, since we call the `commitHash` function anyway, we pass the `cipher` to it at the same time. So storing it on-chain doesn't require making a separate transaction.

Reply:
Also, if your node crashes somehow and you start another one, it can just continue, and won't fail to reveal the secret.
If we stored that locally, the backup node wouldn't have the required information.

Reply:
Would be good to check with some cryptographer if we don't leak some information everytime we encode the same number. I think it's unlikely, but still.

Reply:
You mean because we publish multiple pairs of plaintext + ciphertext for the same key?
Since [ECIES](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme) is supposed to be secure even against chosen-plaintext and chosen-ciphertext attacks, I believe that shouldn't be a problem. But I'm not a cryptographer either.